From f22399508343685e3662828c74182e00b42ac026 Mon Sep 17 00:00:00 2001
Message-ID: <f22399508343685e3662828c74182e00b42ac026.1776946505.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Wed, 8 Apr 2026 17:02:35 +0200
Subject: [RFC PATCH 05/28] memory: add RAMBlockMapOps-aware RAM region init

Add a RAM-backed MemoryRegion initializer that stores provider mapping
callbacks and opaque state before allocating the RAMBlock.

Expose this helper as _nomigrate because provider-backed mappings are
supplied lazily through RAMBlockMapOps callbacks, are host/provider-local,
and are not migration-safe RAM contents.

Later virtio-msg users in this series rely on that non-migratable property
when creating provider-backed mappings.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 include/system/memory.h | 10 ++++++++++
 system/memory.c         | 26 ++++++++++++++++++++++++++
 2 files changed, 36 insertions(+)

diff --git a/include/system/memory.h b/include/system/memory.h
index 0562af313615..3a2bbe1cd53b 100644
--- a/include/system/memory.h
+++ b/include/system/memory.h
@@ -114,6 +114,7 @@ struct MemoryRegionSection {
 };
 
 typedef struct IOMMUTLBEntry IOMMUTLBEntry;
+typedef struct RAMBlockMapOps RAMBlockMapOps;
 
 /*
  * See address_space_translate:
@@ -845,6 +846,8 @@ struct MemoryRegion {
     uint8_t dirty_log_mask;
     bool is_iommu;
     RAMBlock *ram_block;
+    const RAMBlockMapOps *ram_map_ops;
+    void *ram_map_opaque;
     Object *owner;
     /* owner as TYPE_DEVICE. Used for re-entrancy checks in MR access hotpath */
     DeviceState *dev;
@@ -1540,6 +1543,13 @@ void memory_region_init_ram_ptr(MemoryRegion *mr,
                                 const char *name,
                                 uint64_t size,
                                 void *ptr);
+bool memory_region_init_ram_map_nomigrate(MemoryRegion *mr,
+                                          Object *owner,
+                                          const char *name,
+                                          uint64_t size,
+                                          const RAMBlockMapOps *ram_map_ops,
+                                          void *opaque,
+                                          Error **errp);
 
 /**
  * memory_region_init_ram_device_ptr:  Initialize RAM device memory region from
diff --git a/system/memory.c b/system/memory.c
index c51d0798a843..9b4c5a7a48de 100644
--- a/system/memory.c
+++ b/system/memory.c
@@ -1611,6 +1611,32 @@ bool memory_region_init_ram_flags_nomigrate(MemoryRegion *mr,
     return true;
 }
 
+bool memory_region_init_ram_map_nomigrate(MemoryRegion *mr,
+                                          Object *owner,
+                                          const char *name,
+                                          uint64_t size,
+                                          const RAMBlockMapOps *ram_map_ops,
+                                          void *opaque,
+                                          Error **errp)
+{
+    Error *err = NULL;
+
+    memory_region_init(mr, owner, name, size);
+    mr->ram = true;
+    mr->terminates = true;
+    mr->destructor = memory_region_destructor_ram;
+    mr->ram_map_ops = ram_map_ops;
+    mr->ram_map_opaque = opaque;
+    mr->ram_block = qemu_ram_alloc(size, 0, mr, &err);
+    if (err) {
+        mr->size = int128_zero();
+        object_unparent(OBJECT(mr));
+        error_propagate(errp, err);
+        return false;
+    }
+    return true;
+}
+
 bool memory_region_init_resizeable_ram(MemoryRegion *mr,
                                        Object *owner,
                                        const char *name,
-- 
2.53.0

